### ISSUE TITLE
Can not iterate 'string' - ForeachSourceInspector

### ISSUE THREAD

**Issue by funivan** [kalessil/phpinspectionsea#172]
Probably this is false-psitive.
```php
  $files = [
    '1.html',
    '2.html',
  ];
  foreach ($files as $file) { // Get and error
    echo $file;
  }
```

Plugin: `2.3.4`

PhpStorm 2017.1 EAP
Build #PS-171.3691.14, built on March 1, 2017
PhpStorm EAP User
Expiration date: March 31, 2017
JRE: 1.8.0_112-release-724-b6 amd64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Linux 4.4.0-64-generic

**Comment by rentalhost** [comment]
I am running same version than you and I don't get any error (even with master.jar). Are you really running 2.3.4? The ForeachSourceInspector seems to be modified on this version, but on a non-related issue #160.

![image](https://cloud.githubusercontent.com/assets/421851/23657467/01f871a4-031d-11e7-9f3b-1cd7c5dfdb38.png)

**Comment by kalessil** [comment]
#160 applies to function calls only, which are defined in stubs.
I convinced it's due to EAP-related changes,  seems to be a good time to give a try to JB Toolbox =)

**Comment by rentalhost** [comment]
@kalessil Wow... It was very strange... I just closed and opened the PS and I receive the error. haha Still more strange: I closed PS again and no error found.

![image](https://cloud.githubusercontent.com/assets/421851/23657608/abf47bf8-031d-11e7-8876-b4f07a239103.png)

**Comment by kalessil** [comment]
o_O, O_O, x_x

**Comment by rentalhost** [comment]
I never seem that before. I am always running EAP versions, but nothing like that happen at anytime. @funivan Can you select all code, cut and paste on same place? Just to check if the second time will detect some error. On my case, it gones too. Now I don't know in which moment it can happen.

**Comment by kalessil** [comment]
@funivan : can you try reproducing the same behaviour please as @rentalhost ?

**Comment by funivan** [comment]
Yes. It still happen. I just cut/paste/cut etc =) Does not help(

**Comment by funivan** [comment]
I find a way how to reproduce it. This bug affect variables that are defined/used in the global scope.
1. Create empty php project
2. Create file with the following content
```php
 $files = [
    '1.html',
    '2.html',
  ];
  foreach ($files as $file) { // Get and error
    echo $file;
  }
```
At this stage inspection does not fire. Variable `$files` is ok. 
3. Create new php file with following content:
```php
  $files = 'asdf';
```
4. Return to the first file. We get an error.

p.s. I know that variable in global scope are bad practice and i`m working on it =)

**Comment by rjhdby** [comment]
```
    /**
     * @param \SimpleXMLElement $cell
     * @param null|string|string[] $values
     */
    private function setValue(&$cell, $values) {
        if (!is_array($values)) {
            $values = explode(PHP_EOL, $values);
        }
        foreach ($values as $string) {                                 //This string marked as wrong
            $cell->addChild('text:p', $string, OdsDocument::NS_TEXT);
        }
    }
```

> Cannot iterate 'null'...

> Cannot iterate 'string'

Inside foreach $values can be an array only

**Comment by funivan** [comment]
@rjhdby this code is not so good. In php 7 with declare strict types it would not work properly (`$values == null` then you will get error)
Here is some solutions how to resolve your issue:
1. 
```php
   /**
     * @param \SimpleXMLElement $cell
     * @param null|string|string[] $values
     */
    private function setValue(&$cell, $values) {
      $items = [];
      if (is_string($values)) {
        $items = explode(PHP_EOL, $values);
      }
      foreach ($items as $string) {                                 //This string marked as wrong
        $cell->addChild('text:p', $string, OdsDocument::NS_TEXT);
      }
    }
```
2. 
```php
   /**
     * @param \SimpleXMLElement $cell
     * @param null|string|string[] $values
     */
    private function setValue(&$cell, $values) {
      if (!is_array($values)) {
        $values = explode(PHP_EOL, $values);
      }
      foreach ((array) $values as $string) {                                 //This string marked as wrong
        $cell->addChild('text:p', $string, OdsDocument::NS_TEXT);
      }
    }
```

**Comment by Schrank** [comment]
I have the same Problem in Magento. But as in https://github.com/kalessil/phpinspectionsea/issues/172#issuecomment-285271131 described, I think this is intended behaviour. There is no (easy) way to decide wether a file is used in global or non-global context as in 

```
class A {
    public function B() {
        require 'file.php';
    }
}
```
```
//b.php
<?php

// looks like global context
```

**Comment by kalessil** [comment]
Let me check at least why types are so strangely recognized.

**Comment by kalessil** [comment]
Currently, it looks like we have to restrict the scope: functions and methods only.
Then I could reduce a number of false-positives and cover all cases in this report. What do you think guys?

In the worth case, we could drop the inspections - it makes too many problems for quite a while.

**Comment by funivan** [comment]
Agree. Skip check in global scope

**Comment by kalessil** [comment]
On this today, going ahead with checking foreach source only inside functions/methods.

**Comment by kalessil** [comment]
@funivan , @Schrank , @rjhdby : I pushed a fresh binary into the master with fixes and looking for feedback. Does anyone of you want to run the fixed version this and next week?

**Comment by kalessil** [comment]
Closing, will re-open in case of negative feedback.

**Comment by funivan** [comment]
I will try ;)

**Comment by kalessil** [comment]
Great, tests are green but a battlefield is always spots new issues =)

**Comment by kalessil** [comment]
So, it seems like the last changes are good enough, or?

**Comment by kalessil** [comment]
Super =)
